designs/2019-esm-compatibilty/README.md

```diff
@@ -0,0 +1,222 @@
+
+## Drawbacks
+
+### Technical
+
+None. The change has no negative functionality or performance impacts.
+
+## People
+
+Some devs within the Node ecosystem are strongly opposed to supporting `"type": "module"` based packages at all.
+
+## Backwards Compatibility Analysis
+
+*tl;dr: Backward compatibility will be unaffected*
+
+### Story 1 - ESLint Compatibility
+
+This change is additive. It associates `.cjs` files with the existing Javascript configuration loader.
+
+The actual semantics of loading the CommonJS config do not change at all. The `.cjs` extension specifically exists as a escape hatch for this use case (ie loading a CommonJS file that exists in a ESM-based package).
+
+### Story 2 - Node Compatiblity
+
+ESM support rolls out (as a non-experimental feature) w/ Node 13. This includes: the ability to define a package as a ESM; and backward interoperability w/ CJS in ES modules by using the `.cjs` file extension.
+
+This change *only* impacts users who define their package as an ES module. Without Node 13+ that isn't possible.
+
+Existing packages that use the default (ie CommonJS) will be unaffected.
+
+## Alternatives
+
+### Alternative 1 - Suppress the error
+
+**[PR#12333](https://github.com/eslint/eslint/pull/12333)**
+
+There is some debate among the node/modules group surrounding whether the error that causes this issue should be removed altogether.
+
+The fencing exists to prevent users from creating dual-mode (ie both ESM/CJS) packages as interoperability between the two formats has issues. In the case of ESLint, loading a simple config file from a user-defined package should not cause any issues or side-effecs.
+
+Eating the exception is viable solution. Although, this may become dead code if the error is removed from Node's core.
+
+### Alternative 2 - Support dynamic `import()`
+
+Instead of using 'import-fresh' to require the config, import it in a cross-format-compatible way using dynamic `import()`.
+
+There has been and will continue to be a lot of talk about using this approach as the CJS/ESM interop story is fleshed out.
+
+For now it presents too many unknowns and -- even if it didn't -- adding `import()` to ESLint's core would force Node 13+ as a hard constraint.
+
+### Alternative 3: Deprecate JS-based configurations altogether
+
+This approach is controversial but this RFC would be incomplete w/o mentioning it.
+
+What is the benefit of supporting JS-based configurations at all?
+
+- JSON (w/ `strip-comments`) and YAML both support comments
+- JS will load slower than it's context-free alternatives
+- Do the config definitions really require a turing-complete syntax?
+- Is the added complexity of supporting JS configs worth it?
+
+Benefits:
+
+- less complexity
+- reduces ambiguity (ex CJS vs ESM)
+- removes the `import-fresh` dependency
+
+Downsides 
+
+- major breaking change w/ existing patterns
+- JSON requires all strings to be quoted
```

Comment:
JS config files have big advantages than other formats.

- It can switch configuration by environment variables (`process.env`).
- It can reuse other tools configuration. E.g., load `tsconfig.json` and use the `include`/`exclude` setting.
- It can do additional things to plugins. E.g, `require("eslint-plugin-foo").configure({ /*...*/ })`.
- It can use RegExp, bigint, undefined, and functions that JSON doesn't support.

Reply:
I'll take this to mean Alternative 3 isn't a feasible alternative so I removed it from the RFC.